### PR TITLE
Inline fbjs/lib/emptyObject

### PR DESCRIPTION
**Pull request by gaearon**
Continuing https://github.com/facebook/react/pull/13046 and https://github.com/facebook/react/pull/13054.

This one was a bit more challenging, but it was nice to uncover the current assumptions. (Which tended to break with module resetting, for example.)

In most places I just inlined the code to create a DEV-frozen object. But a few places relied on reference equality semantics. I will add inline comments to describe what I did to fix those.

### CODE REVIEW

File: packages/react-reconciler/src/ReactChildFiber.js [hunk 1]
```diff
@@ -157,7 +156,17 @@ function coerceRef(
         return current.ref;
       }
       const ref = function(value) {
-        const refs = inst.refs === emptyObject ? (inst.refs = {}) : inst.refs;
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactChildFiber.js]: This part was a bit shady.

By default we set `inst.refs` to a frozen, shared object. Whenever we want to mutate a ref for the first time, we check whether it's still "the same" pooled object, and if it is, we swap it out with the mutable one.

In some obscure cases where `emptyObject` identity changes over time (e.g. due to a module reset in Jest, or due to a duplicate `fbjs`), this wouldn't work and cause very confusing errors.

I've changed this to use an explicit flag instead. It is only set and read in this module. By default it's not set.

When the flag is not set, the first attempt to set a legacy ref will know that we need to create a new mutable refs object. Then we set the flag. Next time we attach a ref, the flag tells us it's safe to mutate now.

This only affects classes using string refs. There is no cost for classes that don't. Even for those that do, it's a single boolean field.

This lets us get rid of the dependency on strict equality.

File: packages/react-reconciler/src/ReactFiberContext.js [hunk 2]
```diff
@@ -28,14 +27,19 @@ if (__DEV__) {
   warnedAboutMissingGetChildContext = {};
 }
 
+export const emptyContextObject = {};
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberContext.js]: This is exported because there are subtle assumptions around its referential equality in a few places throughout the Fiber codebase. It's used a sentinel value. This makes it explicit for this particular use case.

File: scripts/jest/setupEnvironment.js [hunk 3]
```diff
@@ -46,9 +46,3 @@ if (typeof window !== 'undefined') {
     }
   });
 }
-
-// Preserve the empty object identity across module resets.
-// This is needed for some tests that rely on string refs
-// but reset modules between loading different renderers.
-const obj = require.requireActual('fbjs/lib/emptyObject');
-jest.mock('fbjs/lib/emptyObject', () => obj);
```
Comment (gaearon) [inline scripts/jest/setupEnvironment.js]: We can now drop this.

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.1%, gzip: :small_red_triangle:+0.1%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: ae14317d6810d5b07d01308a2116a707e5e09983...f1ddfdcc9953aa23833b2ce4ddd2390cc4486c66</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | 0.0% | 0.0% | 624.18 KB | 624.43 KB | 145.76 KB | 145.78 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **95.14 KB** | **95.21 KB** | **30.76 KB** | **30.78 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | +0.1% | 616.35 KB | 616.84 KB | 143.57 KB | 143.69 KB | NODE_DEV
**react-dom.production.min.js** | **0.0%** | **:small_red_triangle:+0.1%** | **94.87 KB** | **94.92 KB** | **30.1 KB** | **30.13 KB** | **NODE_PROD**
react-dom-server.browser.development.js | -0.2% | -0.1% | 100.26 KB | 100.04 KB | 26.47 KB | 26.46 KB | UMD_DEV
react-dom-server.browser.development.js | 0.0% | 0.0% | 92.71 KB | 92.72 KB | 24.89 KB | 24.89 KB | NODE_DEV
**react-dom-server.browser.production.min.js** | **-0.2%** | **-0.2%** | **14.41 KB** | **14.38 KB** | **5.52 KB** | **5.5 KB** | **NODE_PROD**
react-dom-server.node.development.js | 0.0% | 0.0% | 94.63 KB | 94.64 KB | 25.42 KB | 25.43 KB | NODE_DEV
**react-dom-server.node.production.min.js** | **-0.2%** | **-0.2%** | **15.21 KB** | **15.19 KB** | **5.82 KB** | **5.81 KB** | **NODE_PROD**
ReactDOM-dev.js | +0.1% | +0.1% | 626.41 KB | 626.89 KB | 142.99 KB | 143.12 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **0.0%** | **:small_red_triangle:+0.1%** | **271.66 KB** | **271.79 KB** | **51.2 KB** | **51.23 KB** | **FB_WWW_PROD**
ReactDOMServer-dev.js | 0.0% | 0.0% | 96.44 KB | 96.45 KB | 24.82 KB | 24.83 KB | FB_WWW_DEV
**ReactDOMServer-prod.js** | **-0.1%** | **-0.1%** | **31.92 KB** | **31.9 KB** | **7.78 KB** | **7.77 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | 0.0% | +0.1% | 95.78 KB | 95.82 KB | 30.4 KB | 30.43 KB | NODE_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.1% | +0.2% | 408.3 KB | 408.57 KB | 91.13 KB | 91.32 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **82.35 KB** | **82.42 KB** | **25.38 KB** | **25.41 KB** | **UMD_PROD**
react-art.development.js | +0.2% | +0.2% | 336.53 KB | 337.04 KB | 73.14 KB | 73.26 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.3%** | **47.07 KB** | **47.12 KB** | **14.61 KB** | **14.66 KB** | **NODE_PROD**
ReactART-dev.js | +0.2% | +0.2% | 329.25 KB | 329.77 KB | 69.06 KB | 69.19 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **142.67 KB** | **142.79 KB** | **24.53 KB** | **24.54 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +0.1% | +0.1% | 342.57 KB | 342.83 KB | 74.17 KB | 74.26 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.4%** | **47.25 KB** | **47.32 KB** | **14.52 KB** | **14.58 KB** | **UMD_PROD**
react-test-renderer.development.js | +0.2% | +0.2% | 334.93 KB | 335.43 KB | 72.09 KB | 72.21 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **46.69 KB** | **46.73 KB** | **14.2 KB** | **14.22 KB** | **NODE_PROD**
react-test-renderer-shallow.development.js | -0.9% | -0.2% | 24.12 KB | 23.89 KB | 6.49 KB | 6.48 KB | UMD_DEV
react-test-renderer-shallow.development.js | 0.0% | +0.3% | 15.69 KB | 15.69 KB | 4.15 KB | 4.16 KB | NODE_DEV
**react-test-renderer-shallow.production.min.js** | **-0.4%** | **-0.2%** | **7.51 KB** | **7.48 KB** | **2.48 KB** | **2.47 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | +0.1% | +0.2% | 340.77 KB | 341.28 KB | 71.7 KB | 71.83 KB | FB_WWW_DEV
ReactShallowRenderer-dev.js | 0.0% | +0.3% | 16.46 KB | 16.47 KB | 4.29 KB | 4.3 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +0.1% | +0.2% | 327.14 KB | 327.63 KB | 69.64 KB | 69.77 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.2%** | **46.8 KB** | **46.84 KB** | **13.98 KB** | **14.01 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | +0.1% | +0.2% | 325.76 KB | 326.24 KB | 69.06 KB | 69.18 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.2%** | **46.81 KB** | **46.85 KB** | **13.99 KB** | **14.01 KB** | **NODE_PROD**

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +0.1% | +0.2% | 460.56 KB | 461.16 KB | 100.99 KB | 101.15 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **206.16 KB** | **206.3 KB** | **36.17 KB** | **36.2 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | +0.1% | +0.2% | 460.27 KB | 460.87 KB | 100.93 KB | 101.09 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **198.72 KB** | **198.86 KB** | **34.72 KB** | **34.75 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | +0.1% | +0.2% | 450.87 KB | 451.4 KB | 98.62 KB | 98.77 KB | RN_FB_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **191 KB** | **191.11 KB** | **33.33 KB** | **33.36 KB** | **RN_FB_PROD**
ReactFabric-dev.js | +0.1% | +0.1% | 450.91 KB | 451.43 KB | 98.63 KB | 98.78 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **191.03 KB** | **191.14 KB** | **33.35 KB** | **33.38 KB** | **RN_OSS_PROD**
ReactNativeRenderer-profiling.js | +0.1% | +0.1% | 201.25 KB | 201.4 KB | 35.27 KB | 35.3 KB | RN_OSS_PROFILING
ReactFabric-profiling.js | +0.1% | +0.1% | 193.25 KB | 193.36 KB | 33.82 KB | 33.84 KB | RN_OSS_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-reconciler/src/ReactChildFiber.js [hunk 4]
```diff
@@ -157,7 +156,17 @@ function coerceRef(
...
+        let refs;
+        // This flag tells us if we need to initialize `this.refs`.
+        if (inst.__reactInternalHasLegacyRefs) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactChildFiber.js]: Small not-even-worth-mentioning nit: since this isn't always a boolean, might be better to explicitly compare to `undefined` instead, like we do for other optional types.

**Comment by sebmarkbage** [comment]
I think we rely on reference equality for empty function too in some cases.

What’s the goal of this? Why not just keep a shared instance?

File: packages/react-reconciler/src/ReactChildFiber.js (hunk 1 above)
Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactChildFiber.js]: You’re messing with the hidden classes here so there can still be a cost.

I’m not sure this is worth it.

File: packages/react-art/src/ReactARTHostConfig.js [hunk 5]
```diff
@@ -9,13 +9,15 @@ import * as ReactScheduler from 'shared/ReactScheduler';
...
 import Mode from 'art/modes/current';
 import invariant from 'fbjs/lib/invariant';
-import emptyObject from 'fbjs/lib/emptyObject';
 
 import {TYPES, EVENT_TYPES, childrenAsString} from './ReactARTInternals';
 
 const pooledTransform = new Transform();
 
-const UPDATE_SIGNAL = {};
+const emptyObject = {};
```
Comment (sebmarkbage) [inline packages/react-art/src/ReactARTHostConfig.js]: Why go from a descript explicit identity to a non-descript, easy to accidentally share object?

Comment (gaearon) [reply packages/react-art/src/ReactARTHostConfig.js]: I was on the fence about this. Even if it's shared, does it matter if it's frozen in DEV? Seems like you'd quickly catch the mistake. It's also used under this name for context. Do you want to have a separate `EMPTY_CONTEXT` too? I can revert that part.

**Comment by gaearon** [comment]
>I think we rely on reference equality for empty function too in some cases.

Don't think so. I've already removed empty function in https://github.com/facebook/react/pull/13054. It wasn't used in comparisons.

>What’s the goal of this? Why not just keep a shared instance but inside the React bundle?

I started doing that but then realized I don't see the point. I also dislike that having it encourages implicit assumptions about reference equality which are later hard to debug.

File: packages/react-reconciler/src/ReactChildFiber.js (hunk 1 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactChildFiber.js]: We're already messing with them for legacy context. The only kind of components I'm worried about regressing (Relay) has both legacy context and legacy refs. This gives an incentive to migrate off both without regressing the current situation.

File: packages/react-art/src/ReactARTHostConfig.js (hunk 5 above)
Comment (gaearon) [reply packages/react-art/src/ReactARTHostConfig.js]: Pushed https://github.com/facebook/react/pull/13055/commits/1f0c930aae723f4a066a8751d1264473bd95b43f which makes explicit naming consistent across renderers

**Comment by sebmarkbage** [comment]
A lot of the legit reasons to use an object is as a tag that cannot be duplicated. Kind of like Symbol. I don’t necessarily see reference equality as a bad thing. It is only bad if two objects that are two conceptual different signals have shared identity.

File: packages/react-reconciler/src/ReactChildFiber.js (hunk 1 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactChildFiber.js]: I guess context is more localized in practice than legacy refs. Maybe not worth risking.

I can keep it on the React object but I think this will require some mocking of React itself in tests to ensure the object is preserved throughout `resetModules`. It was a bit easier for a separate module.

**Comment by gaearon** [comment]
>It is only bad if two objects that are two conceptual different signals have shared identity.

I agree, but that's exactly what happened. "Empty refs" and "empty context" and "empty host context" and "update signal" are all different things. I found it difficult to understand where it was important in the code and where it wasn't. So that's the motivation for not sharing them.

File: packages/react-reconciler/src/ReactChildFiber.js (hunk 1 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactChildFiber.js]: Pushed an alternative fix

**Comment by gaearon** [comment]
Okay, so I pushed an alternative in 2e006cc47596b985b58ad8fa00eaa15c7b5d2791 which doesn't rely on cross-package object identity working, and doesn't break hidden classes. It *does* use `Object.assign` on the object just in case but in practice this shouldn't matter because I expect initial `inst.refs` to always be empty so it shouldn't mess up the hidden class of `LegacyRefsObject` itself. And of course later we set and delete on it anyway.

**Comment by sebmarkbage** [comment]
I see the problem you’re trying to fix now. I was under the impression that we set refs in the renderer and not the isomorphic package.

For everything else we set it up in both. Eg updater can be different until after the constructor.

So what if we do the same here and set it to the renderer’s instance? That way it is always the same.

**Comment by gaearon** [comment]
> I was under the impression that we set refs in the renderer and not the isomorphic package.

We set in both. Normally checking for the renderer-owned `emptyObject` would be sufficient.

I think it's insufficient if the component is *created* by renderer `A` but ends up being an string ref owner of a component rendered by a renderer `B` (e.g. a duplicate `ReactDOM`). I'm not even sure this is possible but maybe it is in some edge case?

Like

```js
ReactDOM.render(<Parent />, root) // parent.refs = emptyObjectFromReactDOM

class Parent {
  render() {
    var el = <Child ref="child" /> // <--- Parent is owner
    return <Indirection>{el}</Indirection>
  }
}

class Indirection {
  componentDidMount() {
    AnotherReactDOMCopy.render(this.props.children, root2)
    // element._owner.refs !== emptyObjectFromAnotherReactDOMCopy
    // so it will think it's safe to assign to them, but they're frozen
  }
}
```

**Comment by sebmarkbage** [comment]
Interesting. If that is possible it is only because they share the current owner. In which case they have a shared isomorphic package.

The reason I’m pushing on this is since we have plans on relying even more on a shared isomorphic package and that I generally think of the OCaml/Rust style of variants to be the appropriate way to handle conditional states. So it’s interesting to explore beyond this change.

I don’t like exposing objects with prototype chains because people inspect them and it gets different hidden classes than other objects. Typically this gets an empty object with expandos.

**Comment by gaearon** [comment]
Wouldn't it switch to dictionary mode anyway as soon as we `delete` on it? (Just a few lines below, when detaching a ref.)

**Comment by gaearon** [comment]
I added a regression test for the case I explained above (which is easy to break with alternative solutions), and pushed a different approach. I'm relying on `React` for sharing the empty object (specifically for refs), and read it one from `new React.Component().refs`. Unlike sharing on a React internal, this will work even for older isomorphic React versions.

**Comment by gaearon** [comment]
The last commit changes the ReactART test because it hits an unfortunate edge case (string refs + resetModules) that breaks the current setup. To be clear it was broken before too, but shimming `emptyObject` in tests worked around it.

A longer term fix might be to extract all shared state (including current owner etc) into yet another package as we discussed at some point, and then always preserving it across module resets, and locking its version.

**Comment by gaearon** [comment]
I'm going to land this. Looks like sharing refs was the only concern and Seb suggested putting it on the React object, I did it through `new React.Component().refs`. Test coverage I added should be helpful if we want to change this again (e.g. via `React.__emptyObject` or something else).
